File: 2025/src/bin/day3/day3_p1.rs
use std::fs;
use std::collections::BinaryHeap;
use std::collections::BTreeMap;

fn main() {

    fn get_heap(line: &str, range: Vec<usize>) -> BinaryHeap<(u32, usize)> {
        let mut h = BinaryHeap::new();
        for (i, num) in line.chars().enumerate() {
            if range.contains(&i) {
                h.push((num.to_digit(10).unwrap(), line.len()-i));
           }
        }
        h
    }

    let contents = fs::read_to_string("src/bin/day3/data.txt")
        .expect("Should have been able to read the file");

    let mut sum = 0;
    let joltage_len = 2;
    for line in contents.lines() {
        let mut joltage_map = BTreeMap::new();
        let mut start = 0;
        let mut end = line.len();
        let mut range: Vec<usize> = (start..end).collect();
        for i in 0..joltage_len {
            let mut h = get_heap(line, range.clone());
            let c = h.pop();
            let k = line.len()-c.unwrap().1;
            let v = c.unwrap().0;
            joltage_map.insert(k ,v);
            if k < range.len()-(joltage_len-(i)) {
                start = k+1;
            } else {
                end = k;
            }
            range = (start..end).collect();
        }
        let joltage_str: String = joltage_map.values().map(|i| i.to_string()).collect::<String>();
        sum += joltage_str.parse::<i32>().unwrap();
        println!("{}, {}, {}", line, joltage_str, sum);
    }
}

File: 2022/day3/day3_p1.py
import string

def main():
    with open("day3/data.txt", "r") as f:
        contents = f.read()

    lines = contents.splitlines()
    result = 0
    priorities = list(string.ascii_lowercase+string.ascii_uppercase)
    for line in lines:
        if len(line)%2 != 0:
            print('odd')
        x = line[0:int(len(line)/2)]
        y = line[int(len(line)/2):]
        for i in x: 
            if i in y:
                result += (priorities.index(i)+1)
                break
                
    print(result)


if __name__ == "__main__":
    main()
